Encrypt bootstrap and direct connections

povilasb opened the pull request:
Basically what it says on the tin.
Also, worth mentioning that currently service discovery test is disabled, because it would be failing otherwise.
This will be fixed once encryption is implemented for service discovery.

povilasb committed:
feat/crypto: add anonymous encryption to crypto context

Anonymous/unauthenticated encryption is helpful when remote peer does not have our public key, but we have theirs.
This scenario will be used for direct and bootstrap connections and service discovery.

povilasb committed:
feat/encryption: encrypt direct connections too

Initial Connect message is sent encrypted with their public key. Connect message includes our public key.
When remote peer receives first Connect message, it extracts public key and starts authenticated encryption aftewards.

povilasb committed:
feat/encryption: encrypt bootstrap connections

Same method as with direct connections used: send initial connection request together with your public key and encrypt it with peer's public key.

povilasb committed:
fix/bootstrap: fix tests

Use proper public keys during bootstrap tests.

maidsafe-highfive commented:
Thanks for the pull request, and welcome! The MaidSafe team is excited to review your changes, and you should hear from @canndrew (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. The way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [CONTRIBUTOR.md](../blob/master/CONTRIBUTOR) for more information.

povilasb committed:
refactor/bootstrap: extract multiple function arguments into structure

povilasb committed:
chore/deps: update secure_serialisation

rust_sodium must be updated too to avoid conflicts with the one secure_serialisation depends on.

canndrew commented on src/crypto.rs:
```diff
@@ -43,58 +44,95 @@ quick_error! {
             description("Error decrypting message")
             display("Error decrypting message: {:?}", e)
         }
+        InvalidOperation(msg: String) {
```
Can either make this take an enum or split it into two separate variants? I don't really like that it carries a string describing the error.

povilasb replied:
sure thing, give me a minute, I'll amend that :)

povilasb committed:
refactor/crypto: use more specific error variants

Allows to easier handle errors.
